// @ulaval/modul-components/dist/components/checkbox/checkbox.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import Component from 'vue-class-component';
import { Emit, Model, Prop, Watch } from 'vue-property-decorator';
import { InputState } from '../../mixins/input-state/input-state';
import uuid from '../../utils/uuid/uuid';
import { ModulVue } from '../../utils/vue/vue';
import { CHECKBOX_NAME } from '../component-names';
import ValidationMessagePlugin from '../validation-message/validation-message';
import WithRender from './checkbox.html?style=./checkbox.scss';
export var MCheckboxPosition;
(function (MCheckboxPosition) {
    MCheckboxPosition["Left"] = "left";
    MCheckboxPosition["Right"] = "right";
})(MCheckboxPosition || (MCheckboxPosition = {}));
var MCheckbox = /** @class */ (function (_super) {
    __extends(MCheckbox, _super);
    function MCheckbox() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.isFocus = false;
        _this.id = "mCheckbox-" + uuid.generate();
        _this.internalValue = false;
        return _this;
    }
    MCheckbox.prototype.onChange = function (value) { };
    MCheckbox.prototype.onClick = function (event) {
        this.$refs['checkbox']['blur']();
    };
    MCheckbox.prototype.onValueChange = function (value) {
        this.internalValue = value;
    };
    Object.defineProperty(MCheckbox.prototype, "propValue", {
        get: function () {
            return this.value !== undefined ? this.value : this.internalValue;
        },
        set: function (value) {
            this.onChange(value);
            this.internalValue = value;
        },
        enumerable: true,
        configurable: true
    });
    MCheckbox.prototype.setFocus = function (value) {
        this.isFocus = value;
    };
    Object.defineProperty(MCheckbox.prototype, "propIndeterminate", {
        get: function () {
            return this.indeterminate && !this.propValue;
        },
        set: function (newValue) {
            this.indeterminate = newValue;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MCheckbox.prototype, "hasCheckboxLeft", {
        get: function () {
            return this.position === MCheckboxPosition.Left;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MCheckbox.prototype, "hasLabelSlot", {
        get: function () {
            return !!this.$slots.default;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MCheckbox.prototype, "forId", {
        get: function () {
            if (this.as().readonly) {
                return undefined;
            }
            else {
                return this.id;
            }
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        Model('change'),
        Prop()
    ], MCheckbox.prototype, "value", void 0);
    __decorate([
        Prop({ default: false })
    ], MCheckbox.prototype, "indeterminate", void 0);
    __decorate([
        Prop({
            default: MCheckboxPosition.Left,
            validator: function (value) {
                return value === MCheckboxPosition.Left ||
                    value === MCheckboxPosition.Right;
            }
        })
    ], MCheckbox.prototype, "position", void 0);
    __decorate([
        Emit('change')
    ], MCheckbox.prototype, "onChange", null);
    __decorate([
        Emit('click')
    ], MCheckbox.prototype, "onClick", null);
    __decorate([
        Watch('value')
    ], MCheckbox.prototype, "onValueChange", null);
    MCheckbox = __decorate([
        WithRender,
        Component({
            mixins: [InputState]
        })
    ], MCheckbox);
    return MCheckbox;
}(ModulVue));
export { MCheckbox };
var CheckboxPlugin = {
    install: function (v, options) {
        v.prototype.$log.debug(CHECKBOX_NAME, 'plugin.install');
        v.use(ValidationMessagePlugin);
        v.component(CHECKBOX_NAME, MCheckbox);
    }
};
export default CheckboxPlugin;
