// @ulaval/modul-components/dist/components/radio-style/radio-style.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';
import { ModulVue } from '../../utils/vue/vue';
import { RADIO_STYLE_NAME } from '../component-names';
import WithRender from './radio-style.html?style=./radio-style.scss';
export var MRadioStylePosition;
(function (MRadioStylePosition) {
    MRadioStylePosition["Left"] = "left";
    MRadioStylePosition["Right"] = "right";
})(MRadioStylePosition || (MRadioStylePosition = {}));
var MRadioStyle = /** @class */ (function (_super) {
    __extends(MRadioStyle, _super);
    function MRadioStyle() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(MRadioStyle.prototype, "isInputRight", {
        get: function () {
            return this.radioPosition === MRadioStylePosition.Right;
        },
        enumerable: true,
        configurable: true
    });
    MRadioStyle.prototype.onClick = function (event) {
        if (!this.disabled) {
            this.$emit('click', event);
        }
    };
    __decorate([
        Prop({
            default: MRadioStylePosition.Right,
            validator: function (value) {
                return value === MRadioStylePosition.Left ||
                    value === MRadioStylePosition.Right;
            }
        })
    ], MRadioStyle.prototype, "radioPosition", void 0);
    __decorate([
        Prop({ default: '1em' })
    ], MRadioStyle.prototype, "inputSize", void 0);
    __decorate([
        Prop()
    ], MRadioStyle.prototype, "focus", void 0);
    __decorate([
        Prop()
    ], MRadioStyle.prototype, "checked", void 0);
    __decorate([
        Prop()
    ], MRadioStyle.prototype, "disabled", void 0);
    __decorate([
        Prop()
    ], MRadioStyle.prototype, "fullWidth", void 0);
    MRadioStyle = __decorate([
        WithRender,
        Component
    ], MRadioStyle);
    return MRadioStyle;
}(ModulVue));
export { MRadioStyle };
var RadioStylePlugin = {
    install: function (v, options) {
        v.component(RADIO_STYLE_NAME, MRadioStyle);
    }
};
export default RadioStylePlugin;
