// @ulaval/modul-components/dist/components/plus/plus.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';
import { ModulVue } from '../../utils/vue/vue';
import { PLUS_NAME } from '../component-names';
import WithRender from './plus.html?style=./plus.scss';
export var MPlusSkin;
(function (MPlusSkin) {
    MPlusSkin["Default"] = "default";
    MPlusSkin["Light"] = "light";
    MPlusSkin["CurrentColor"] = "current-color";
})(MPlusSkin || (MPlusSkin = {}));
var MPlus = /** @class */ (function (_super) {
    __extends(MPlus, _super);
    function MPlus() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MPlus.prototype.onClick = function (event) {
        if (!this.disabled) {
            this.$emit('click', event);
        }
    };
    __decorate([
        Prop()
    ], MPlus.prototype, "open", void 0);
    __decorate([
        Prop()
    ], MPlus.prototype, "large", void 0);
    __decorate([
        Prop()
    ], MPlus.prototype, "border", void 0);
    __decorate([
        Prop({
            default: MPlusSkin.Default,
            validator: function (value) {
                return value === MPlusSkin.Default ||
                    value === MPlusSkin.Light ||
                    value === MPlusSkin.CurrentColor;
            }
        })
    ], MPlus.prototype, "skin", void 0);
    __decorate([
        Prop()
    ], MPlus.prototype, "disabled", void 0);
    MPlus = __decorate([
        WithRender,
        Component
    ], MPlus);
    return MPlus;
}(ModulVue));
export { MPlus };
var PlusPlugin = {
    install: function (v, options) {
        v.component(PLUS_NAME, MPlus);
    }
};
export default PlusPlugin;
