// @ulaval/modul-components/dist/mixins/input-width/input-width.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';
import { ModulVue } from '../../utils/vue/vue';
export var InputMaxWidth;
(function (InputMaxWidth) {
    InputMaxWidth["None"] = "none";
    InputMaxWidth["XSmall"] = "x-small";
    InputMaxWidth["Small"] = "small";
    InputMaxWidth["Regular"] = "regular";
    InputMaxWidth["Medium"] = "medium";
    InputMaxWidth["Large"] = "large";
})(InputMaxWidth || (InputMaxWidth = {}));
var InputWidth = /** @class */ (function (_super) {
    __extends(InputWidth, _super);
    function InputWidth() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(InputWidth.prototype, "inputMaxWidth", {
        get: function () {
            var maxWidth;
            switch (this.maxWidth) {
                case InputMaxWidth.None:
                    maxWidth = undefined;
                    break;
                case InputMaxWidth.XSmall:
                    maxWidth = '72px';
                    break;
                case InputMaxWidth.Small:
                    maxWidth = '144px';
                    break;
                case InputMaxWidth.Regular:
                    maxWidth = '288px';
                    break;
                case InputMaxWidth.Medium:
                    maxWidth = '608px';
                    break;
                case InputMaxWidth.Large:
                    maxWidth = '800px';
                    break;
                default:
                    maxWidth = this.maxWidth;
            }
            return maxWidth;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputWidth.prototype, "inputWidth", {
        get: function () {
            return this.width === 'none' ? undefined : this.width;
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        Prop({ default: '100%' })
    ], InputWidth.prototype, "width", void 0);
    __decorate([
        Prop({ default: InputMaxWidth.Regular })
    ], InputWidth.prototype, "maxWidth", void 0);
    InputWidth = __decorate([
        Component
    ], InputWidth);
    return InputWidth;
}(ModulVue));
export { InputWidth };
