// @ulaval/modul-components/dist/components/component-names.js
export var ACCORDION_GROUP_NAME = 'm-accordion-group';
export var ACCORDION_NAME = 'm-accordion';
export var ACCORDION_TRANSITION_NAME = 'm-accordion-transition';
export var BUTTON_NAME = 'm-button';
export var BUTTON_GROUP_NAME = 'm-button-group';
export var CALENDAR_NAME = 'm-calendar';
export var CAROUSEL_NAME = 'm-carousel';
export var CAROUSEL_ITEM_NAME = 'm-carousel-item';
export var CHARACTER_COUNT_NAME = 'm-character-count';
export var CHECKBOX_NAME = 'm-checkbox';
export var DATEFIELDS_NAME = 'm-datefields';
export var DATEPICKER_NAME = 'm-datepicker';
export var DATERANGEPICKER_NAME = 'm-daterangepicker';
export var DIALOG_NAME = 'm-dialog';
export var DROPDOWN_NAME = 'm-dropdown';
export var DROPDOWN_ITEM_NAME = 'm-dropdown-item';
export var DROPDOWN_GROUP_NAME = 'm-dropdown-group';
export var DYNAMIC_TEMPLATE_NAME = 'm-dynamic-template';
export var ERROR_ACCESS_DENIED_NAME = 'm-error-access-denied';
export var ERROR_BROWSER_NOT_SUPPORTED_NAME = 'm-error-browser-not-supported';
export var ERROR_CONFIG_NOT_SUPPORTED_NAME = 'm-error-config-not-supported';
export var ERROR_COOKIES_NOT_SUPPORTED_NAME = 'm-error-cookies-not-supported';
export var ERROR_MESSAGE_NAME = 'm-error-message';
export var ERROR_PAGE_NOT_FOUND_NAME = 'm-error-page-not-found';
export var ERROR_TECHNICAL_DIFFICULTY_NAME = 'm-error-technical-difficulty';
export var FILE_SELECT_NAME = 'm-file-select';
export var FILE_UPLOAD_NAME = 'm-file-upload';
export var FORM = 'm-form';
export var OVERLAY_NAME = 'm-overlay';
export var FLEX_TEMPLATE_NAME = 'm-flex-template';
export var I18N_NAME = 'm-i18n';
export var ICON_NAME = 'm-icon';
export var ICON_BUTTON_NAME = 'm-icon-button';
export var ICON_FILE_NAME = 'm-icon-file';
export var INPLACE_EDIT_NAME = 'm-inplace-edit';
export var INPUT_GROUP_NAME = 'm-input-group';
export var INPUT_STYLE_NAME = 'm-input-style';
export var LIMIT_TEXT_NAME = 'm-limit-text';
export var LINK_NAME = 'm-link';
export var LIST_ITEM_NAME = 'm-list-item';
export var MESSAGE_NAME = 'm-message';
export var MENU_NAME = 'm-menu';
export var MENU_ITEM_NAME = 'm-menu-item';
export var MESSAGE_PAGE_NAME = 'm-message-page';
export var MODAL_NAME = 'm-modal';
export var NAVBAR_NAME = 'm-navbar';
export var NAVBAR_ITEM_NAME = 'm-navbar-item';
export var OPTION_NAME = 'm-option';
export var OPTION_ITEM_NAME = 'm-option-item';
export var OPTION_ITEM_ADD_NAME = 'm-option-item-add';
export var OPTION_ITEM_ARCHIVE_NAME = 'm-option-item-archive';
export var OPTION_ITEM_DELETE_NAME = 'm-option-item-delete';
export var OPTION_ITEM_EDIT_NAME = 'm-option-item-edit';
export var PAGE_NOT_FOUND_NAME = 'm-page-not-found';
export var PANEL_NAME = 'm-panel';
export var PERIODPICKER_NAME = 'm-periodpicker';
export var PHONE_NUMBER_NAME = 'm-phone-number';
export var PLUS_NAME = 'm-plus';
export var POPPER_NAME = 'm-popper';
export var POPUP_NAME = 'm-popup';
export var PROGRESS_NAME = 'm-progress';
export var RADIO_NAME = 'm-radio';
export var RADIO_GROUP_NAME = 'm-radio-group';
export var RADIO_STYLE_NAME = 'm-radio-style';
export var RICH_TEXT_NAME = 'm-rich-text';
export var RICH_TEXT_EDITOR_NAME = 'm-rich-text-editor';
export var SCROLL_TOP_NAME = 'm-scroll-top';
export var SESSION_EXPIRED_NAME = 'm-session-expired';
export var SHOW_MORE_NAME = 'm-show-more';
export var SIDEBAR_NAME = 'm-sidebar';
export var SLIDER_NAME = 'm-slider';
export var SPINNER_NAME = 'm-spinner';
export var STATUS_NAME = 'm-status';
export var STEP_NAME = 'm-step';
export var STEPPERS_NAME = 'm-steppers';
export var STEPPERS_ITEM_NAME = 'm-steppers-item';
export var SWITCH_NAME = 'm-switch';
export var TAB_PANEL_NAME = 'm-tab-panel';
export var TABLE_NAME = 'm-table';
export var TABLE_HEADER_NAME = 'm-table-header';
export var TABS_NAME = 'm-tabs';
export var TEMPLATE_NAME = 'm-template';
export var TEXTAREA_NAME = 'm-textarea';
export var TEXTFIELD_NAME = 'm-textfield';
export var TIMEPICKER_NAME = 'm-timepicker';
export var TOAST = 'm-toast';
export var TOOLTIP_NAME = 'm-tooltip';
export var TOUCH_NAME = 'm-touch';
export var TREE_NAME = 'm-tree';
export var TREE_ICON_NAME = 'm-tree-icon';
export var TREE_NAME_NODE = 'm-tree-node';
export var VALIDATION_MESSAGE_NAME = 'm-validation-message';
