// @ulaval/modul-components/dist/components/calendar/calendar-renderer/abstract-calendar-renderer.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import { Emit, Prop } from 'vue-property-decorator';
import { ModulVue } from '../../../utils/vue/vue';
import { CalendarEvent } from '../calendar-state/state/calendar-state';
var MAbstractCalendarRenderer = /** @class */ (function (_super) {
    __extends(MAbstractCalendarRenderer, _super);
    function MAbstractCalendarRenderer() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MAbstractCalendarRenderer.prototype.onDaySelect = function (day) {
    };
    MAbstractCalendarRenderer.prototype.onDayMouseEnter = function (day) {
    };
    MAbstractCalendarRenderer.prototype.onDayMouseLeave = function (day) {
    };
    MAbstractCalendarRenderer.prototype.onMonthSelect = function (month) {
    };
    MAbstractCalendarRenderer.prototype.onMonthNext = function (event) {
    };
    MAbstractCalendarRenderer.prototype.onMonthPrevious = function (event) {
    };
    MAbstractCalendarRenderer.prototype.onYearSelect = function (year) {
    };
    MAbstractCalendarRenderer.prototype.onYearNext = function (event) {
    };
    MAbstractCalendarRenderer.prototype.onYearPrevious = function (event) {
    };
    __decorate([
        Prop({ required: true })
    ], MAbstractCalendarRenderer.prototype, "calendar", void 0);
    __decorate([
        Emit(CalendarEvent.DAY_SELECT)
    ], MAbstractCalendarRenderer.prototype, "onDaySelect", null);
    __decorate([
        Emit(CalendarEvent.DAY_MOUSE_ENTER)
    ], MAbstractCalendarRenderer.prototype, "onDayMouseEnter", null);
    __decorate([
        Emit(CalendarEvent.DAY_MOUSE_LEAVE)
    ], MAbstractCalendarRenderer.prototype, "onDayMouseLeave", null);
    __decorate([
        Emit(CalendarEvent.MONTH_SELECT)
    ], MAbstractCalendarRenderer.prototype, "onMonthSelect", null);
    __decorate([
        Emit(CalendarEvent.MONTH_NEXT)
    ], MAbstractCalendarRenderer.prototype, "onMonthNext", null);
    __decorate([
        Emit(CalendarEvent.MONTH_PREVIOUS)
    ], MAbstractCalendarRenderer.prototype, "onMonthPrevious", null);
    __decorate([
        Emit(CalendarEvent.YEAR_SELECT)
    ], MAbstractCalendarRenderer.prototype, "onYearSelect", null);
    __decorate([
        Emit(CalendarEvent.YEAR_NEXT)
    ], MAbstractCalendarRenderer.prototype, "onYearNext", null);
    __decorate([
        Emit(CalendarEvent.YEAR_PREVIOUS)
    ], MAbstractCalendarRenderer.prototype, "onYearPrevious", null);
    return MAbstractCalendarRenderer;
}(ModulVue));
export { MAbstractCalendarRenderer };
