// @ulaval/modul-components/dist/directives/scroll-to/scroll-to.js
import Vue from 'vue';
import { ScrollToDuration } from '../../utils/scroll-to/scroll-to';
import { SCROLL_TO_NAME } from '../directive-names';
var ScrollToCallback = /** @class */ (function () {
    function ScrollToCallback(speed, offset, target) {
        var _this = this;
        this.speed = speed;
        this.offset = offset;
        this.target = target;
        this.callBack = function (event) {
            var scrollTo = Vue.prototype.$scrollTo;
            scrollTo.goTo(_this.target, _this.offset, _this.speed);
        };
    }
    return ScrollToCallback;
}());
var MScrollTo = {
    inserted: function (element, binding, node) {
        var speed = binding.value.speed || ScrollToDuration.Regular;
        var offset = binding.value.offset || 0;
        if (!node.context) {
            throw new Error('Error node context is null');
        }
        var target = node.context.$refs[binding.arg];
        var _scrollToCallback = new ScrollToCallback(speed, offset, target);
        Object.defineProperty(element, '_scrollToCallback', {
            value: _scrollToCallback
        });
        element.addEventListener('touchstart', _scrollToCallback.callBack);
        element.addEventListener('click', _scrollToCallback.callBack);
    },
    update: function (element, binding) {
        if (element && element._scrollToCallback) {
            element._scrollToCallback.speed = binding.value.speed || ScrollToDuration.Regular;
            element._scrollToCallback.offset = binding.value.offset || 0;
        }
    },
    unbind: function (element, binding) {
        if (element && element._scrollToCallback) {
            element.removeEventListener('touchstart', element._scrollToCallback.callBack);
            element.removeEventListener('click', element._scrollToCallback.callBack);
        }
    }
};
var ScrollToPlugin = {
    install: function (v, options) {
        v.prototype.$log.error('ScrollToDirective will be deprecated in modul v.1.0');
        v.use(ScrollToPlugin);
        v.directive(SCROLL_TO_NAME, MScrollTo);
    }
};
export default ScrollToPlugin;
