// @ulaval/modul-components/dist/mixins/input-state/input-state.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';
import { ModulVue } from '../../utils/vue/vue';
export var InputStateValue;
(function (InputStateValue) {
    InputStateValue["Default"] = "default";
    InputStateValue["Disabled"] = "disabled";
    InputStateValue["Waiting"] = "waiting";
    InputStateValue["Error"] = "error";
    InputStateValue["Valid"] = "valid";
})(InputStateValue || (InputStateValue = {}));
export var InputStateTagStyle;
(function (InputStateTagStyle) {
    InputStateTagStyle["Default"] = "default";
    InputStateTagStyle["H1"] = "h1";
    InputStateTagStyle["H2"] = "h2";
    InputStateTagStyle["H3"] = "h3";
    InputStateTagStyle["H4"] = "h4";
    InputStateTagStyle["H5"] = "h5";
    InputStateTagStyle["H6"] = "h6";
    InputStateTagStyle["P"] = "p";
})(InputStateTagStyle || (InputStateTagStyle = {}));
var InputState = /** @class */ (function (_super) {
    __extends(InputState, _super);
    function InputState() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    Object.defineProperty(InputState.prototype, "active", {
        get: function () {
            return !this.isDisabled && !this.isWaiting && !this.readonly;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputState.prototype, "isDisabled", {
        get: function () {
            return this.state === InputStateValue.Disabled;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputState.prototype, "isWaiting", {
        get: function () {
            return this.state === InputStateValue.Waiting;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputState.prototype, "hasError", {
        get: function () {
            return this.state === InputStateValue.Error;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputState.prototype, "isValid", {
        get: function () {
            return this.state === InputStateValue.Valid;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputState.prototype, "state", {
        get: function () {
            var state;
            if (!this.disabled) {
                if (!this.waiting) {
                    if (this.hasErrorMessage || this.error) {
                        state = InputStateValue.Error;
                    }
                    else if (this.hasValidMessage || this.valid) {
                        state = InputStateValue.Valid;
                    }
                    else {
                        state = InputStateValue.Default;
                    }
                }
                else {
                    state = InputStateValue.Waiting;
                }
            }
            else {
                state = InputStateValue.Disabled;
            }
            return state;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputState.prototype, "hasErrorMessage", {
        get: function () {
            return (!!this.errorMessage || this.errorMessage === ' ') && !this.disabled && !this.waiting;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputState.prototype, "hasValidMessage", {
        get: function () {
            return (!!this.validMessage || this.validMessage === ' ') && !this.disabled && !this.waiting && !this.hasErrorMessage;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputState.prototype, "hasHelperMessage", {
        get: function () {
            return (!!this.helperMessage || this.helperMessage === ' ') && !this.disabled && !this.waiting;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(InputState.prototype, "hasValidationMessage", {
        get: function () {
            return (this.hasErrorMessage || this.hasValidMessage || this.hasHelperMessage) && this.active;
        },
        enumerable: true,
        configurable: true
    });
    InputState.prototype.getInput = function () {
        var selector = this.as().selector || 'input, textarea, [contenteditable=true]';
        var elements = this.$el.querySelectorAll(selector);
        if (elements.length > 1) {
            throw new Error("Input state can manage 1 and only 1 nested editable element (" + selector + ")");
        }
        return elements[0];
    };
    __decorate([
        Prop()
    ], InputState.prototype, "disabled", void 0);
    __decorate([
        Prop()
    ], InputState.prototype, "waiting", void 0);
    __decorate([
        Prop()
    ], InputState.prototype, "error", void 0);
    __decorate([
        Prop()
    ], InputState.prototype, "valid", void 0);
    __decorate([
        Prop()
    ], InputState.prototype, "errorMessage", void 0);
    __decorate([
        Prop()
    ], InputState.prototype, "validMessage", void 0);
    __decorate([
        Prop()
    ], InputState.prototype, "helperMessage", void 0);
    __decorate([
        Prop({
            default: InputStateTagStyle.Default,
            validator: function (value) {
                return value === InputStateTagStyle.Default ||
                    value === InputStateTagStyle.H1 ||
                    value === InputStateTagStyle.H2 ||
                    value === InputStateTagStyle.H3 ||
                    value === InputStateTagStyle.H4 ||
                    value === InputStateTagStyle.H5 ||
                    value === InputStateTagStyle.H6 ||
                    value === InputStateTagStyle.P;
            }
        })
    ], InputState.prototype, "tagStyle", void 0);
    __decorate([
        Prop()
    ], InputState.prototype, "readonly", void 0);
    InputState = __decorate([
        Component
    ], InputState);
    return InputState;
}(ModulVue));
export { InputState };
