// @ulaval/modul-components/dist/components/accordion/accordion-transition.js
import Vue from 'vue';
import { ACCORDION_TRANSITION_NAME } from '../component-names';
export var MAccordionTransition = Vue.extend({
    functional: true,
    render: function (createElement, context) {
        var props = context.props;
        var CLASS_HAS_TRANSITION = 'm-accordion--has-transition';
        var data = {
            props: {
                name: 'm-accordion'
            },
            on: {
                beforeEnter: function (el) {
                    if (props.transition || props.transition === undefined) {
                        el.classList.add(CLASS_HAS_TRANSITION);
                    }
                    else if (el.classList.contains(CLASS_HAS_TRANSITION)) {
                        el.classList.remove(CLASS_HAS_TRANSITION);
                    }
                },
                enter: function (el) {
                    var heightDelta = props.heightDelta
                        ? props.heightDelta
                        : 0;
                    el.style.height = el.scrollHeight - heightDelta + 'px';
                },
                afterEnter: function (el) {
                    el.style.removeProperty('height');
                },
                beforeLeave: function (el) {
                    el.style.height = parseInt(window.getComputedStyle(el).height, 10) + 'px';
                    if (props.transition === false && el.classList.contains(CLASS_HAS_TRANSITION)) {
                        el.classList.remove(CLASS_HAS_TRANSITION);
                    }
                },
                afterLeave: function (el) {
                    el.style.removeProperty('height');
                }
            }
        };
        return createElement('transition', data, context.children);
    }
});
var AccordionTransitionPlugin = {
    install: function (v, options) {
        v.component(ACCORDION_TRANSITION_NAME, MAccordionTransition);
    }
};
export default AccordionTransitionPlugin;
