// @ulaval/modul-components/dist/components/file-select/file-select.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';
import { InputState } from '../../mixins/input-state/input-state';
import FilePlugin, { DEFAULT_STORE_NAME } from '../../utils/file/file';
import uuid from '../../utils/uuid/uuid';
import { ModulVue } from '../../utils/vue/vue';
import ButtonPlugin, { MButtonIconPosition, MButtonSkin } from '../button/button';
import { FILE_SELECT_NAME } from '../component-names';
import I18nPlugin from '../i18n/i18n';
import ValidationMesagePlugin from '../validation-message/validation-message';
import WithRender from './file-select.html?style=./file-select.scss';
var MFileSelect = /** @class */ (function (_super) {
    __extends(MFileSelect, _super);
    function MFileSelect() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.id = "mFileSelect-" + uuid.generate();
        return _this;
    }
    MFileSelect.prototype.destroyed = function () {
        if (!this.keepStore) {
            this.$file.destroy(this.storeName);
        }
    };
    MFileSelect.prototype.onClick = function (event) {
        this.$refs.inputFile.click();
        this.$emit('click', event);
        this.$refs.inputFile.blur();
    };
    MFileSelect.prototype.onFocus = function (event) {
        this.$emit('focus');
    };
    MFileSelect.prototype.onBlur = function (event) {
        this.$emit('blur');
    };
    MFileSelect.prototype.processFile = function (event) {
        var file = this.$refs.inputFile.files;
        if (file) {
            this.$file.add(file, this.storeName);
        }
        this.$refs.inputFile.value = '';
    };
    Object.defineProperty(MFileSelect.prototype, "hasLabel", {
        get: function () {
            return !!this.label;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MFileSelect.prototype, "extensions", {
        get: function () {
            return this.allowedExtensions.map(function (ext) { return (ext.startsWith('.') ? '' : '.') + ext; }).join(', ');
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        Prop()
    ], MFileSelect.prototype, "label", void 0);
    __decorate([
        Prop({
            default: MButtonSkin.Secondary,
            validator: function (value) {
                return value === MButtonSkin.Primary || value === MButtonSkin.Secondary;
            }
        })
    ], MFileSelect.prototype, "skin", void 0);
    __decorate([
        Prop()
    ], MFileSelect.prototype, "disabled", void 0);
    __decorate([
        Prop()
    ], MFileSelect.prototype, "waiting", void 0);
    __decorate([
        Prop()
    ], MFileSelect.prototype, "fullSize", void 0);
    __decorate([
        Prop()
    ], MFileSelect.prototype, "iconName", void 0);
    __decorate([
        Prop({
            default: MButtonIconPosition.Left,
            validator: function (value) {
                return value === MButtonIconPosition.Left ||
                    value === MButtonIconPosition.Right;
            }
        })
    ], MFileSelect.prototype, "iconPosition", void 0);
    __decorate([
        Prop({ default: '12px' })
    ], MFileSelect.prototype, "iconSize", void 0);
    __decorate([
        Prop()
    ], MFileSelect.prototype, "multiple", void 0);
    __decorate([
        Prop({
            default: DEFAULT_STORE_NAME
        })
    ], MFileSelect.prototype, "storeName", void 0);
    __decorate([
        Prop({ default: false })
    ], MFileSelect.prototype, "keepStore", void 0);
    __decorate([
        Prop({ default: function () { return []; } })
    ], MFileSelect.prototype, "allowedExtensions", void 0);
    MFileSelect = __decorate([
        WithRender,
        Component({
            mixins: [InputState]
        })
    ], MFileSelect);
    return MFileSelect;
}(ModulVue));
export { MFileSelect };
var FileSelectPlugin = {
    install: function (v, options) {
        v.use(I18nPlugin);
        v.use(ButtonPlugin);
        v.use(ValidationMesagePlugin);
        v.use(FilePlugin);
        v.component(FILE_SELECT_NAME, MFileSelect);
    }
};
export default FileSelectPlugin;
