// @ulaval/modul-components/dist/mixins/input-popup/input-popup.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import Component from 'vue-class-component';
import { KeyCode } from '../../utils/keycode/keycode';
import { ModulVue } from '../../utils/vue/vue';
import { InputState } from '../input-state/input-state';
import { MediaQueries } from '../media-queries/media-queries';
var InputPopup = /** @class */ (function (_super) {
    __extends(InputPopup, _super);
    function InputPopup() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.internalValue = '';
        return _this;
    }
    InputPopup.prototype.inputOnKeydownEnter = function ($event) {
        if (!this.open) {
            this.open = true;
        }
        $event.preventDefault();
    };
    InputPopup.prototype.inputOnKeydownTab = function ($event) {
        if (this.as().isMqMinS) {
            this.open = false;
        }
    };
    InputPopup.prototype.inputOnKeydownDelete = function () {
        this.internalValue = '';
    };
    InputPopup.prototype.inputOnKeydown = function ($event) {
        // tslint:disable: deprecation
        if ($event.keyCode !== KeyCode.M_RETURN &&
            $event.keyCode !== KeyCode.M_ENTER &&
            $event.keyCode !== KeyCode.M_TAB &&
            $event.keyCode !== KeyCode.M_SHIFT &&
            $event.keyCode !== KeyCode.M_ESCAPE && !this.open) {
            this.open = true;
        }
        // tslint:enable: deprecation
    };
    InputPopup.prototype.inputOnFocus = function () {
        if (!this.open && !this.as().isDisabled && this.as().isMqMinS) {
            this.open = true;
        }
    };
    Object.defineProperty(InputPopup.prototype, "isEmpty", {
        get: function () {
            return this.as().hasValue || (this.hasPlaceholder() && this.open) ? false : true;
        },
        enumerable: true,
        configurable: true
    });
    InputPopup.prototype.hasPlaceholder = function () {
        return this.placeholder !== undefined && this.placeholder !== '';
    };
    InputPopup = __decorate([
        Component({
            mixins: [
                InputState,
                MediaQueries
            ]
        })
    ], InputPopup);
    return InputPopup;
}(ModulVue));
export { InputPopup };
