// @ulaval/modul-components/dist/components/radio-group/radio-group.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import Component from 'vue-class-component';
import { Emit, Model, Prop, Watch } from 'vue-property-decorator';
import { InputState } from '../../mixins/input-state/input-state';
import uuid from '../../utils/uuid/uuid';
import { RADIO_GROUP_NAME } from '../component-names';
import RadioPlugin, { BaseRadioGroup, MRadioPosition, MRadioVerticalAlignement } from '../radio/radio';
import ValidationMessagePlugin from '../validation-message/validation-message';
import WithRender from './radio-group.html?style=./radio-group.scss';
var MRadioGroup = /** @class */ (function (_super) {
    __extends(MRadioGroup, _super);
    function MRadioGroup() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.name = uuid.generate();
        _this.internalValue = '';
        return _this;
    }
    MRadioGroup.prototype.onChange = function (value) { };
    MRadioGroup.prototype.onFocus = function (event) { };
    MRadioGroup.prototype.onBlur = function (event) { };
    Object.defineProperty(MRadioGroup.prototype, "stateIsDisabled", {
        get: function () {
            return this.as().isDisabled;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MRadioGroup.prototype, "stateIsError", {
        get: function () {
            return this.as().hasError;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MRadioGroup.prototype, "stateIsValid", {
        get: function () {
            return this.as().isValid;
        },
        enumerable: true,
        configurable: true
    });
    MRadioGroup.prototype.getValue = function () {
        return this.model;
    };
    MRadioGroup.prototype.updateValue = function (value) {
        this.model = value;
    };
    MRadioGroup.prototype.created = function () {
        this.internalValue = undefined;
    };
    MRadioGroup.prototype.onValueChange = function (value) {
        this.internalValue = value;
    };
    Object.defineProperty(MRadioGroup.prototype, "model", {
        get: function () {
            return this.value === undefined ? this.internalValue : this.value;
        },
        set: function (value) {
            this.internalValue = value;
            this.onChange(value);
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MRadioGroup.prototype, "hasLabel", {
        get: function () {
            return !!this.label;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MRadioGroup.prototype, "idLabel", {
        get: function () {
            return this.hasLabel ? uuid.generate() : undefined;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MRadioGroup.prototype, "idValidationMessage", {
        get: function () {
            return this.as().errorMessage || this.as().validMessage || this.as().helperMessage ? uuid.generate() : undefined;
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        Model('change'),
        Prop()
    ], MRadioGroup.prototype, "value", void 0);
    __decorate([
        Prop({
            default: MRadioPosition.Left,
            validator: function (value) {
                return value === MRadioPosition.Left ||
                    value === MRadioPosition.Right;
            }
        })
    ], MRadioGroup.prototype, "radiosPosition", void 0);
    __decorate([
        Prop()
    ], MRadioGroup.prototype, "inline", void 0);
    __decorate([
        Prop()
    ], MRadioGroup.prototype, "label", void 0);
    __decorate([
        Prop({
            default: MRadioVerticalAlignement.Top,
            validator: function (value) {
                return value === MRadioVerticalAlignement.Top ||
                    value === MRadioVerticalAlignement.Center;
            }
        })
    ], MRadioGroup.prototype, "radiosVerticalAlign", void 0);
    __decorate([
        Prop()
    ], MRadioGroup.prototype, "radiosMarginTop", void 0);
    __decorate([
        Prop({
            default: false
        })
    ], MRadioGroup.prototype, "readOnly", void 0);
    __decorate([
        Emit('change')
    ], MRadioGroup.prototype, "onChange", null);
    __decorate([
        Emit('focus')
    ], MRadioGroup.prototype, "onFocus", null);
    __decorate([
        Emit('blur')
    ], MRadioGroup.prototype, "onBlur", null);
    __decorate([
        Watch('value')
    ], MRadioGroup.prototype, "onValueChange", null);
    MRadioGroup = __decorate([
        WithRender,
        Component({
            mixins: [InputState]
        })
    ], MRadioGroup);
    return MRadioGroup;
}(BaseRadioGroup));
export { MRadioGroup };
var RadioGroupPlugin = {
    install: function (v, options) {
        v.use(RadioPlugin);
        v.use(ValidationMessagePlugin);
        v.component(RADIO_GROUP_NAME, MRadioGroup);
    }
};
export default RadioGroupPlugin;
