// @ulaval/modul-components/dist/components/status/status.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import Vue from 'vue';
import Component from 'vue-class-component';
import { Prop } from 'vue-property-decorator';
import { STATUS_NAME } from '../component-names';
import IconPlugin from '../icon/icon';
import WithRender from './status.html?style=./status.scss';
export var MStatusListStates;
(function (MStatusListStates) {
    MStatusListStates["Completed"] = "completed";
    MStatusListStates["Pending"] = "pending";
    MStatusListStates["Error"] = "error";
})(MStatusListStates || (MStatusListStates = {}));
export var MStatusListIconName;
(function (MStatusListIconName) {
    MStatusListIconName["Check"] = "m-svg__completed-filled";
    MStatusListIconName["Error"] = "m-svg__error-filled";
})(MStatusListIconName || (MStatusListIconName = {}));
var MStatus = /** @class */ (function (_super) {
    __extends(MStatus, _super);
    function MStatus() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.statusEnum = MStatusListStates;
        return _this;
    }
    MStatus.prototype.getIconName = function (state) {
        return state === MStatusListStates.Error ? MStatusListIconName.Error : MStatusListIconName.Check;
    };
    Object.defineProperty(MStatus.prototype, "isCompleted", {
        get: function () {
            return this.status === this.statusEnum.Completed;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MStatus.prototype, "isPending", {
        get: function () {
            return this.status === this.statusEnum.Pending;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MStatus.prototype, "isError", {
        get: function () {
            return this.status === this.statusEnum.Error;
        },
        enumerable: true,
        configurable: true
    });
    __decorate([
        Prop({
            validator: function (value) {
                return value === MStatusListStates.Completed ||
                    value === MStatusListStates.Pending ||
                    value === MStatusListStates.Error;
            }
        })
    ], MStatus.prototype, "status", void 0);
    MStatus = __decorate([
        WithRender,
        Component
    ], MStatus);
    return MStatus;
}(Vue));
export { MStatus };
var Status = {
    install: function (v, options) {
        v.prototype.$log.error('MStatus will be deprecated in modul v.1.0');
        v.use(IconPlugin);
        v.component(STATUS_NAME, MStatus);
    }
};
export default Status;
