// @ulaval/modul-components/dist/components/steppers/steppers-item/steppers-item.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import Component from 'vue-class-component';
import { Prop, Watch } from 'vue-property-decorator';
import { ModulVue } from '../../../utils/vue/vue';
import WithRender from './steppers-item.html?style=./steppers-item.scss';
export var MSteppersItemState;
(function (MSteppersItemState) {
    MSteppersItemState["Visited"] = "visited";
    MSteppersItemState["InProgress"] = "in-progress";
    MSteppersItemState["Disabled"] = "disabled";
})(MSteppersItemState || (MSteppersItemState = {}));
var BaseSteppers = /** @class */ (function (_super) {
    __extends(BaseSteppers, _super);
    function BaseSteppers() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    return BaseSteppers;
}(ModulVue));
export { BaseSteppers };
var MSteppersItem = /** @class */ (function (_super) {
    __extends(MSteppersItem, _super);
    function MSteppersItem() {
        return _super !== null && _super.apply(this, arguments) || this;
    }
    MSteppersItem.prototype.stateChanged = function (value) {
        if (this.$parent instanceof BaseSteppers) {
            this.$parent.setAnim(true);
            this.$parent.setLineWidth();
            this.$parent.setAnim(false);
        }
        this.$emit('update:value', this.state);
    };
    Object.defineProperty(MSteppersItem.prototype, "isVisited", {
        get: function () {
            return this.state === MSteppersItemState.Visited;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MSteppersItem.prototype, "isInProgress", {
        get: function () {
            return this.state === MSteppersItemState.InProgress;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MSteppersItem.prototype, "isDisabled", {
        get: function () {
            return this.state === MSteppersItemState.Disabled;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MSteppersItem.prototype, "isTabIndex", {
        get: function () {
            return this.isVisited ? 0 : -1;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MSteppersItem.prototype, "isLink", {
        get: function () {
            return !!this.url && this.isVisited;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MSteppersItem.prototype, "componentToShow", {
        get: function () {
            return this.isLink ? 'router-link' : 'div';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MSteppersItem.prototype, "role", {
        get: function () {
            return this.isLink || this.isDisabled ? undefined : 'button';
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MSteppersItem.prototype, "propUrl", {
        get: function () {
            return this.isLink ? this.url : undefined;
        },
        enumerable: true,
        configurable: true
    });
    MSteppersItem.prototype.onClick = function (event) {
        if (this.isVisited) {
            this.$emit('click', event);
            this.$refs.title.blur();
        }
    };
    __decorate([
        Prop({
            default: MSteppersItemState.Disabled,
            validator: function (value) {
                return value === MSteppersItemState.Visited ||
                    value === MSteppersItemState.InProgress ||
                    value === MSteppersItemState.Disabled;
            }
        })
    ], MSteppersItem.prototype, "state", void 0);
    __decorate([
        Prop()
    ], MSteppersItem.prototype, "iconName", void 0);
    __decorate([
        Prop()
    ], MSteppersItem.prototype, "iconTitle", void 0);
    __decorate([
        Prop({ default: false })
    ], MSteppersItem.prototype, "completed", void 0);
    __decorate([
        Prop()
    ], MSteppersItem.prototype, "url", void 0);
    __decorate([
        Watch('state')
    ], MSteppersItem.prototype, "stateChanged", null);
    MSteppersItem = __decorate([
        WithRender,
        Component
    ], MSteppersItem);
    return MSteppersItem;
}(ModulVue));
export { MSteppersItem };
