// @ulaval/modul-components/dist/components/flex-template/flex-template.js
var __extends = (this && this.__extends) || (function () {
    var extendStatics = function (d, b) {
        extendStatics = Object.setPrototypeOf ||
            ({ __proto__: [] } instanceof Array && function (d, b) { d.__proto__ = b; }) ||
            function (d, b) { for (var p in b) if (b.hasOwnProperty(p)) d[p] = b[p]; };
        return extendStatics(d, b);
    };
    return function (d, b) {
        extendStatics(d, b);
        function __() { this.constructor = d; }
        d.prototype = b === null ? Object.create(b) : (__.prototype = b.prototype, new __());
    };
})();
var __decorate = (this && this.__decorate) || function (decorators, target, key, desc) {
    var c = arguments.length, r = c < 3 ? target : desc === null ? desc = Object.getOwnPropertyDescriptor(target, key) : desc, d;
    if (typeof Reflect === "object" && typeof Reflect.decorate === "function") r = Reflect.decorate(decorators, target, key, desc);
    else for (var i = decorators.length - 1; i >= 0; i--) if (d = decorators[i]) r = (c < 3 ? d(r) : c > 3 ? d(target, key, r) : d(target, key)) || r;
    return c > 3 && r && Object.defineProperty(target, key, r), r;
};
import Component from 'vue-class-component';
import { Prop, Watch } from 'vue-property-decorator';
import { ElementQueries } from '../../mixins/element-queries/element-queries';
import ModulPlugin from '../../utils/modul/modul';
import { ModulVue } from '../../utils/vue/vue';
import { FLEX_TEMPLATE_NAME } from '../component-names';
import WithRender from './flex-template.html?style=./flex-template.scss';
export var MFlexTemplateOrigin;
(function (MFlexTemplateOrigin) {
    MFlexTemplateOrigin["Left"] = "left";
    MFlexTemplateOrigin["Right"] = "right";
})(MFlexTemplateOrigin || (MFlexTemplateOrigin = {}));
var MFlexTemplate = /** @class */ (function (_super) {
    __extends(MFlexTemplate, _super);
    function MFlexTemplate() {
        var _this = _super !== null && _super.apply(this, arguments) || this;
        _this.menuOpenCount = 0;
        _this.transitionDelayOpen = false;
        _this.paddingPage = '';
        _this.topMenu = '';
        _this.headerHeight = 0;
        _this.headerHidden = false;
        _this.scrollPosition = 0;
        _this.internalMenuOpen = false;
        _this.menuHasAnim = false;
        _this.animDynamicHeader = false;
        return _this;
    }
    MFlexTemplate.prototype.mounted = function () {
        var _this = this;
        this.propMenuOpen = this.menuOpen;
        this.scrollPosition = this.$modul.scrollPosition;
        this.$on('isEqMaxS', function (value) { return _this.isEqMaxSChanged(value); });
        this.setHeaderHeight();
        this.$modul.event.$on('scroll', this.onScroll);
        this.$modul.event.$on('resizeDone', this.onResizeDone);
    };
    MFlexTemplate.prototype.beforeDestroy = function () {
        this.$modul.event.$off('scroll', this.onScroll);
        this.$modul.event.$off('resizeDone', this.onResizeDone);
    };
    MFlexTemplate.prototype.menuOpenChanged = function (open) {
        this.menuHasAnim = true;
        this.propMenuOpen = open;
        if (open) {
            this.$emit('open');
        }
        else {
            this.$emit('close');
        }
    };
    MFlexTemplate.prototype.smallMenuChanged = function (smallMenu) {
        this.menuHasAnim = true;
        if (this.propMenuOpen && !this.as().isEqMaxS) {
            this.setMenuWidth();
        }
    };
    MFlexTemplate.prototype.isEqMaxSChanged = function (value) {
        if (this.propMenuOpen) {
            this.setMenuWidth();
        }
    };
    MFlexTemplate.prototype.onScroll = function () {
        this.adjustFixeMenu();
        this.adjustDynamicHeader();
    };
    MFlexTemplate.prototype.onResizeDone = function (e) {
        this.setHeaderHeight();
        this.adjustFixeMenu();
        this.adjustDynamicHeader();
    };
    MFlexTemplate.prototype.setHeaderHeight = function () {
        var _this = this;
        this.$nextTick(function () {
            if (_this.hasHeaderSlot) {
                _this.headerHeight = _this.$refs.header.clientHeight;
                _this.setSpacing(_this.headerHeight);
            }
        });
    };
    MFlexTemplate.prototype.setSpacing = function (spacing) {
        this.paddingPage = this.propHeaderFixe ? spacing + 'px' : '';
        this.topMenu = ((this.propHeaderFixe || this.propMenuFixe) && !this.isMenuFixeFake) || this.as().isEqMaxS ? spacing + 'px' : '';
    };
    MFlexTemplate.prototype.adjustDynamicHeader = function () {
        if (this.propDynamicHeader) {
            var position = this.$el.getBoundingClientRect().top;
            var maxPosition = position + this.headerHeight + 50;
            this.headerHidden = (maxPosition <= 0 && !this.propMenuOpen) && (this.scrollPosition >= position);
            this.scrollPosition = position;
        }
    };
    MFlexTemplate.prototype.adjustFixeMenu = function () {
        var _this = this;
        if (!this.propHeaderFixe && this.propMenuFixe && this.propMenuOpen && !this.as().isEqMaxS) {
            var menuContainer = this.$refs.menuContainer;
            var menu = this.$refs.menu;
            var topPosition_1 = menuContainer.getBoundingClientRect().top;
            setTimeout(function () {
                if (topPosition_1 <= 0) {
                    _this.topMenu = Math.abs(topPosition_1) + 'px';
                }
                else {
                    _this.topMenu = '';
                }
            }, 60);
        }
    };
    MFlexTemplate.prototype.setMenuWidth = function () {
        var width = this.propSmallMenu ? this.smallMenuWidth : this.menuWidth;
        var menuContainer = this.$refs.menuContainer;
        var menu = this.$refs.menu;
        var pageContainer = this.$refs.pageContainer;
        menuContainer.style.width = width;
        menu.style.width = width;
        pageContainer.style.width = 'calc(100% - ' + width + ')';
    };
    Object.defineProperty(MFlexTemplate.prototype, "propMenuOpen", {
        get: function () {
            return this.internalMenuOpen;
        },
        set: function (open) {
            this.internalMenuOpen = this.hasMenuSlot ? open : false;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MFlexTemplate.prototype, "propSmallMenu", {
        get: function () {
            return this.hasMenuSlot ? this.smallMenu : false;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MFlexTemplate.prototype, "propHeaderFixe", {
        get: function () {
            return this.hasHeaderSlot && this.headerFixe;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MFlexTemplate.prototype, "propMenuFixe", {
        get: function () {
            return this.hasMenuSlot && this.menuFixe;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MFlexTemplate.prototype, "propDynamicHeader", {
        get: function () {
            return this.dynamicHeader && this.propHeaderFixe;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MFlexTemplate.prototype, "propPageMinHeight", {
        get: function () {
            return this.pageMinHeight === undefined || this.pageMinHeight === '' ? '100vh' : this.pageMinHeight;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MFlexTemplate.prototype, "isMenuOriginRight", {
        get: function () {
            return this.menuOrigin === MFlexTemplateOrigin.Right;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MFlexTemplate.prototype, "isMenuFixeFake", {
        get: function () {
            return !this.propHeaderFixe && this.propMenuFixe;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MFlexTemplate.prototype, "hasHeaderSlot", {
        get: function () {
            return !!this.$slots.header;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MFlexTemplate.prototype, "hasMenuSlot", {
        get: function () {
            return !!this.$slots.menu;
        },
        enumerable: true,
        configurable: true
    });
    Object.defineProperty(MFlexTemplate.prototype, "hasFooterSlot", {
        get: function () {
            return !!this.$slots.footer;
        },
        enumerable: true,
        configurable: true
    });
    MFlexTemplate.prototype.transitionEnter = function (el, done) {
        var _this = this;
        this.adjustFixeMenu();
        this.adjustDynamicHeader();
        if (!this.as().isEqMaxS) {
            if (this.menuHasAnim) {
                setTimeout(function () {
                    _this.transitionDelayOpen = true;
                    _this.setMenuWidth();
                    setTimeout(function () {
                        done();
                    }, 450);
                }, 20);
            }
            else {
                this.transitionDelayOpen = true;
                this.setMenuWidth();
                done();
            }
        }
        else {
            setTimeout(function () {
                _this.transitionDelayOpen = true;
            }, 20);
            done();
        }
    };
    MFlexTemplate.prototype.transitionAfterEnter = function (el) {
        var menuContent = this.$refs.menuContent;
        if (this.menuOpenCount !== 0) {
            menuContent.focus();
        }
        this.menuOpenCount++;
        // tslint:disable-next-line: deprecation
        this.$modul.updateAfterResize();
    };
    MFlexTemplate.prototype.transitionLeave = function (el, done) {
        var _this = this;
        this.transitionDelayOpen = false;
        if (!this.as().isEqMaxS) {
            var menuContainer = this.$refs.menuContainer;
            var pageContainer = this.$refs.pageContainer;
            menuContainer.style.removeProperty('width');
            pageContainer.style.removeProperty('width');
        }
        setTimeout(function () {
            // tslint:disable-next-line: deprecation
            _this.$modul.updateAfterResize();
            done();
        }, 450);
    };
    __decorate([
        Prop({ default: true })
    ], MFlexTemplate.prototype, "dynamicHeader", void 0);
    __decorate([
        Prop({ default: true })
    ], MFlexTemplate.prototype, "headerFixe", void 0);
    __decorate([
        Prop({ default: '300px' })
    ], MFlexTemplate.prototype, "menuWidth", void 0);
    __decorate([
        Prop({ default: true })
    ], MFlexTemplate.prototype, "menuFixe", void 0);
    __decorate([
        Prop({
            default: MFlexTemplateOrigin.Left,
            validator: function (value) {
                return value === MFlexTemplateOrigin.Left ||
                    value === MFlexTemplateOrigin.Right;
            }
        })
    ], MFlexTemplate.prototype, "menuOrigin", void 0);
    __decorate([
        Prop()
    ], MFlexTemplate.prototype, "menuOpen", void 0);
    __decorate([
        Prop()
    ], MFlexTemplate.prototype, "smallMenu", void 0);
    __decorate([
        Prop({ default: '44px' })
    ], MFlexTemplate.prototype, "smallMenuWidth", void 0);
    __decorate([
        Prop()
    ], MFlexTemplate.prototype, "pageMinHeight", void 0);
    __decorate([
        Watch('menuOpen')
    ], MFlexTemplate.prototype, "menuOpenChanged", null);
    __decorate([
        Watch('smallMenu')
    ], MFlexTemplate.prototype, "smallMenuChanged", null);
    MFlexTemplate = __decorate([
        WithRender,
        Component({
            mixins: [ElementQueries]
        })
    ], MFlexTemplate);
    return MFlexTemplate;
}(ModulVue));
export { MFlexTemplate };
var FlexTemplatePlugin = {
    install: function (v, options) {
        v.prototype.$log.error('MFlexTemplate will be deprecated in modul v.1.0');
        v.use(ModulPlugin);
        v.component(FLEX_TEMPLATE_NAME, MFlexTemplate);
    }
};
export default FlexTemplatePlugin;
